12 - 記事投稿機能

Dimda opened the pull request:
## 説明
- 記事投稿機能
- 記事投稿機能自動テスト
- リファクタリング

## Issue 
#12

ReiOogami committed:
12 - 記事投稿機能追加

問題
記事の投稿機能が無かった

解決策
package.json に validator を追加
投稿のためのActions,Reducer,Container,Componentを作成
Router に投稿画面へのルートを追加
LoginFormとRegistrationFormのsnapshotを変更

Commit:
12 - 記事投稿の自動テスト

問題
記事投稿機能がテストされていない。

解決策
テストに必要なfakerを追加と過去で追加するのを忘れたjestようのeslintパケージ追加。
firebaseに'database'やpushなどのモック機能を追加とリファクタリング。
react-native-router-fluxにpopのモック機能を追加。
記事投稿に関連する Action creator, Reducer, Componentのテストを作成。

Commit:
12 - error component 作成

問題
複数の箇所で同じコードが繰り返し使用されていた

解決策
エラーメッセージを表示するcomponentを作成した

Commit:
12 - リファクタリング

問題
記事投稿機能で使っていた書き方と今まで使っていた書き方に違いがある。

解決策
今まで書いたコンポーネントを軽くリファクタリングする。
要らないテストケースを削除。

Dimda committed:
12 - ログインの時にスピナー表示

問題
ログインの時にスピナーが表示されない。

解決策
ログイン処理が始まる時に'AUTH_PROCESSING'がdispatchされてなかったので追加した。
テストケースを更新。

youmjww commented on src/actions/ArticleActions.js:
```diff
@@ -0,0 +1,49 @@
+  if (validator.isInt(plantName) || plantName >= 25 || plantName === '') {
+    return {
+      type: ADD_ARTICLE_FAIL,
+      payload: '植物名は文字列かつ、25文字以下で入力してください。'
```
25文字のタイトルをつけるとエラーが出た
日本語では以下は含まれるので、この場合、
`植物名は文字列かつ、25文字未満で入力してください。`
か
`植物名は文字列かつ、24文字以下で入力してください。`
が適当だと思われる

youmjww commented on src/actions/ArticleActions.js:
```diff
@@ -0,0 +1,49 @@
+  if (validator.isInt(articleContents) || articleContents >= 1000 || articleContents === '') {
+    return {
+      type: ADD_ARTICLE_FAIL,
+      payload: '記事内容は文字列かつ、1000文字以下で入力してください。'
```
上と同様1000文字ちょうどだとエラーが出る

70mas commented:
記事投稿の写真は別チケット？

70mas reviewed (commented):
記事投稿の写真は別チケット？

70mas commented:
投稿するときの確認はないんですか

youmjww commented on __tests__/actionCreators/ArticleActions.test.js:
```diff
@@ -0,0 +1,79 @@
+
+  it('1000文字以上の記事を入力して失敗する', () => {
+    const plantName = 'ひまわり';
+    const articleContents = faker.lorem.words(1000);
```
ここで特定の文字数のダミーテキストを出力してくれるメソッドを使っているのなら、39行目も`faker.lorem.words();`で合わせて欲しい

Dimda replied:
直した

youmjww commented on src/actions/ArticleActions.js:
```diff
@@ -23,13 +23,13 @@ export const saveArticle = (plantName, articleContents) => {
   const { uid } = currentUser;
 
-  if (validator.isInt(plantName) || plantName >= 25 || plantName === '') {
+  if (validator.isInt(plantName) || plantName.length >= 25 || plantName === '') {
```
マジックナンバー・・・

Dimda replied:
よく変わってる値じゃないと思うからそのままでいいと思う。

youmjww commented on src/actions/ArticleActions.js:
```diff
@@ -23,13 +23,13 @@ export const saveArticle = (plantName, articleContents) => {
     };
   }
-  if (validator.isInt(articleContents) || articleContents >= 1000 || articleContents === '') {
+  if (validator.isInt(articleContents) || articleContents.length >= 1000 || articleContents === '') {
```
ここもマジックナンバー

Dimda replied:
よく変わってる値じゃないと思うからそのままでいいと思う。

70mas reviewed (dismissed):
ok

youmjww commented on src/components/AddArticle.js:
```diff
@@ -0,0 +1,70 @@
+            height={120}
+          />
+        </CardSection>
+        <Text style={styles.errorTextStyle}>
```
ここに改行が欲しい

youmjww commented on __tests__/components/__snapshots__/LoginForm.test.js.snap:
```diff
@@ -332,6 +339,7 @@ exports[`ログインフォーム snapshot-メールアドレスとパスワー
             "paddingRight": 5,
           }
         }
+        underlineColorAndroid="transparent"
```
`transparent`だと読みづらいので、キャメルケースかスネークケースか、文字列なのでスペースを入れて欲しい（他も同様）

ReiOogami replied:
react-nativeの要素なので変更不可能

youmjww commented on src/components/AddArticle.js:
```diff
@@ -0,0 +1,58 @@
+            onChangeText={this.onArticleContentsChanged.bind(this)}
+            value={this.props.articleName}
+            multiline={true}
+            height={120}
```
マジックナンバー
特にこの数字は他のところでもよく使いそうなので定数にしたい

Dimda replied:
もうInputの要素で`height`が固定されているから、この場合は上書きしている。

youmjww commented on __tests__/reducers/ArticleReducer.test.js:
```diff
@@ -0,0 +1,69 @@
+    expect(reducer(MODIFIED_STATE, {
+      type: types.ADD_ARTICLE_FAIL,
+      payload: 'error',
+    })).toEqual({ 
```
trailing spaceがある

youmjww commented on src/actions/ArticleActions.js:
```diff
@@ -0,0 +1,49 @@
+  ADD_ARTICLE_SUCCESS,
+  ADD_ARTICLE_FAIL
+} from './types';
+import firebase from 'firebase';
```
node_modules系は上に持ってきて、独自のimportは下に持って行きたい

Dimda commented:
@70mas うん、写真と確認は別のチケットになる。

youmjww reviewed (changes requested):
以上気になる点です。
確認お願いします。

youmjww commented:
<img width="294" alt="2017-10-24 14 41 49" src="https://user-images.githubusercontent.com/13462428/31926480-9cb98d60-b8c9-11e7-83d2-bf1249671c89.png">
issueカードと実際に格納されているレコードの順番が違うけど、問題ない？

Dimda commented:
@youmjww アルファベット順になってるので大丈夫と思う。でも時期投稿の日時フィールドが将来的にあった方がいい。

ReiOogami committed:
12 - レビューでの問題点を変更

問題
レビューで指摘された点があった。

解決策
fakerと使用するように統一
tralingspaceを削除
importの順番を変更
エラーメッセージの数字を修正
addarticleで使用されているcomponentの間に改行を入れる

youmjww commented on __tests__/actionCreators/ArticleActions.test.js:
```diff
@@ -36,7 +36,7 @@ describe('記事関連のアクション', () => {
 
   it('25文字以上の植物名を入力して失敗する', () => {
-    const plantName = 'ひまわりひまわりひまわりひまわりひまわりひまわりひまわりひまわりひまわりひまわりひま';
+    const plantName = faker.lorem.words(10);
```
25文字以上じゃなくて10文字でテスト？

youmjww commented on __tests__/actionCreators/ArticleActions.test.js:
```diff
@@ -0,0 +1,79 @@
+    expect(actions.articleContentsChanged(articleContents)).toEqual(expectedAction);
+  });
+
+  it('記事内容の変更', () => {
```
一個上のテストとの差分がわからないけど、同じことしてる？

youmjww commented on __tests__/actionCreators/ArticleActions.test.js:
```diff
@@ -36,7 +36,7 @@ describe('記事関連のアクション', () => {
   });
 
   it('25文字以上の植物名を入力して失敗する', () => {
```
ここのテストの名前は`26文字以上`にしておいたほうがいいと思う

youmjww commented on __tests__/actionCreators/ArticleActions.test.js:
```diff
@@ -0,0 +1,79 @@
+    expect(actions.saveArticle(plantName, articleContents)).toEqual(expectedAction);
+  });
+
+  it('1000文字以上の記事を入力して失敗する', () => {
```
ここのテストのタイトルも`1001文字以上`にしたほうがいいと思う(てか、なんでテストが通っているんだろ・・・)

youmjww reviewed (changes requested):
気になった点は以上です。
確認よろしくお願いします。

youmjww commented:
問題ないと思うのでマージします。
